File: C++/2-TwoPointers/Algorithms/1-Fast-and-SlowPointers(Floyd’sCycleDetection)/Fast-and-SlowPointersPro.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <memory> // For smart pointers

// 1. Linked List Cycle Detection (Classic Floyd's Algorithm)
//
//   * Real-world application: Detecting loops in linked lists representing data structures, such as detecting if a linked list representing a queue or stack has been corrupted and has an infinite loop.
//   * Explanation: This is the classic application of the Floyd's algorithm.  A linked list node contains a value and a pointer to the next node.  A cycle exists if, starting from the head, you can reach the same node again by following the 'next' pointers.
//   * Time Complexity: O(n)
//   * Space Complexity: O(1)
//
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

bool hasCycle(ListNode* head) {
    if (!head || !head->next) {
        return false;
    }
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) {
            return true;
        }
    }
    return false;
}

// 2. Detecting Cycles in Object Relationships
//
//   * Real-world application: Detecting circular dependencies in object graphs, such as in a dependency injection container or a scene graph in a game engine.  This prevents infinite loops during object construction or updates.
//   * Explanation:  Imagine objects where one object holds a reference to another.  A cycle exists if object A references B, B references C, and C references A.  This can cause problems in garbage collection or object serialization.  We can simulate this with a simplified object structure.  We use unique_ptr to manage the memory.
//   * Time Complexity: O(n) - where n is the number of objects.
//   * Space Complexity: O(1)
//
struct ObjectNode {
    int id;
    std::unique_ptr<ObjectNode> reference; // Using unique_ptr to manage ownership
    ObjectNode(int i) : id(i), reference(nullptr) {}
};

bool hasCircularReference(ObjectNode* startNode) {
    if (!startNode) return false;

    ObjectNode* slow = startNode;
    ObjectNode* fast = startNode;

    while (fast && fast->reference) {
        slow = slow->reference.get(); // Use .get() to get the raw pointer
        fast = fast->reference->reference.get(); // Access nested reference, safely.
        if (slow == fast) {
            return true;
        }
    }
    return false;
}

// 3. Finding the Start of a Cycle in a Linked List
//
//   * Real-world application:  Identifying the entry point of a loop in data structures, which is crucial for debugging memory leaks or data corruption issues in complex systems.
//   * Explanation:  This extends the basic cycle detection to find the *starting* node of the cycle.  This is important for understanding the cause of the cycle.
//   * Time Complexity: O(n)
//   * Space Complexity: O(1)
//
ListNode* findCycleStart(ListNode* head) {
    if (!head || !head->next) {
        return nullptr;
    }
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) { // Cycle detected
            slow = head;
            while (slow != fast) {
                slow = slow->next;
                fast = fast->next;
            }
            return slow; // Returns the start of the cycle
        }
    }
    return nullptr; // No cycle
}

// 4. Detecting Cycles in a Finite State Machine (FSM)
//
//  * Real-world application: Verifying the correctness of state transitions in a finite state machine, used in embedded systems, protocol design, or game development.  A cycle might indicate a state machine that gets stuck in a loop.
//  * Explanation: A finite state machine consists of states and transitions between them.  We can model states as nodes and transitions as pointers.  A cycle means the FSM can return to a previous state, which might be intended or a bug.
//  * Time Complexity: O(n), where n is the number of states visited.
//  * Space Complexity: O(1)
//
enum class FSMState {
    STATE_A,
    STATE_B,
    STATE_C,
    STATE_D,
    STATE_E,
    STATE_F
};

FSMState getNextState(FSMState currentState, std::vector<FSMState>& history) {
    history.push_back(currentState); // Keep track of visited states
    switch (currentState) {
        case FSMState::STATE_A: return FSMState::STATE_B;
        case FSMState::STATE_B: return FSMState::STATE_C;
        case FSMState::STATE_C: return FSMState::STATE_D;
        case FSMState::STATE_D: return FSMState::STATE_E;
        case FSMState::STATE_E: return FSMState::STATE_F;
        case FSMState::STATE_F: return FSMState::STATE_B; // Cycle back to B
        default: return FSMState::STATE_A;
    }
}

bool hasCycleFSM() {
    FSMState slow = FSMState::STATE_A;
    FSMState fast = FSMState::STATE_A;
    std::vector<FSMState> history;

    for (int i = 0; i < 10; ++i) { // Limit iterations to prevent infinite loop in case of error
        slow = getNextState(slow, history);
        fast = getNextState(getNextState(fast, history), history);
        if (slow == fast) {
            return true;
        }
    }
    return false;
}

// 5. Detecting Cycles in Iterators
//
//   * Real-world application:  Analyzing iterative processes where the same state can be reached multiple times, such as in simulations or game AI. Detecting cycles can help identify repeating patterns or potential infinite loops in the logic.
//   * Explanation: This applies the concept to any process where you repeatedly apply a function to get the next state.  The "next state" is like the 'next' pointer in a linked list.  We simulate this with a simple integer iterator.
//   * Time Complexity: O(n), where n is the number of iterations.
//   * Space Complexity: O(1)
//
int nextValue(int currentValue, std::vector<int>& history) {
    history.push_back(currentValue);
    return (currentValue * 2 + 1) % 10; // Example iteration function
}

bool hasCycleIterator() {
    int slow = 0;
    int fast = 0;
    std::vector<int> history;
    for (int i = 0; i < 20; ++i) { // Limit iterations
        slow = nextValue(slow, history);
        fast = nextValue(nextValue(fast, history), history);
        if (slow == fast) {
            return true;
        }
    }
    return false;
}

int main() {
    // 1. Linked List Cycle Detection
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = head->next; // Create a cycle
    std::cout << "Linked List has cycle: " << (hasCycle(head) ? "true" : "false") << std::endl; // Output: true

    // Clean up linked list
    ListNode* current = head;
    ListNode* next;
    while (current) {
        next = current->next;
        // IMPORTANT:  Do NOT delete the node that closes the cycle (head->next in this case) twice.
        if (current != head->next) {
           delete current;
        }
        current = next;
        if (current == head) break; // Stop at the start of the cycle.
    }
    //  delete head->next; //  <- DON'T DO THIS.  It was already deleted.
    head = nullptr; // Set head to nullptr to prevent dangling pointer

    // 2. Detecting Cycles in Object Relationships
    std::unique_ptr<ObjectNode> obj1 = std::make_unique<ObjectNode>(1);
    std::unique_ptr<ObjectNode> obj2 = std::make_unique<ObjectNode>(2);
    std::unique_ptr<ObjectNode> obj3 = std::make_unique<ObjectNode>(3);

    obj1->reference = std::move(obj2);
    obj2->reference = std::move(obj3);
    obj3->reference = std::move(obj1); // Create a cycle
    std::cout << "Object graph has circular reference: " << (hasCircularReference(obj1.get()) ? "true" : "false") << std::endl; // Output: true

    // 3. Finding the Start of a Cycle in a Linked List
     head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = head->next; // Create a cycle
    ListNode* cycleStart = findCycleStart(head);
    if (cycleStart) {
        std::cout << "Cycle starts at node with value: " << cycleStart->val << std::endl; // Output: 2
    } else {
        std::cout << "No cycle found" << std::endl;
    }
     // Clean up linked list
     current = head;
     while (current) {
        next = current->next;
        if (current != head->next){
             delete current;
        }
        current = next;
        if (current == head) break;
    }
    head = nullptr;

    // 4. Detecting Cycles in a Finite State Machine
    std::cout << "FSM has cycle: " << (hasCycleFSM() ? "true" : "false") << std::endl; // Output: true

    // 5. Detecting Cycles in Iterators
    std::cout << "Iterator has cycle: " << (hasCycleIterator() ? "true" : "false") << std::endl; // Output: true

    return 0;
}

File: Java/2-TwoPointers/Algorithms/1-Fast-and-SlowPointers(Floyd’sCycleDetection)/Fast-and-SlowPointersPro.java
import java.util.*;

public class TwoPointers {

    // 1. Linked List Cycle Detection
    //   - Real-world application: Detecting loops in linked list-based data structures,
    //     such as detecting if a routing path in a network has a loop, which can cause network congestion.
    //   - Problem: Given the head of a singly linked list, determine if the list has a cycle.
    //   - Time Complexity: O(n)
    //   - Space Complexity: O(1)
    public static boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head.next; // Start fast one step ahead

        while (fast != null && fast.next != null) {
            if (slow == fast) {
                return true; // Cycle detected
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        return false; // No cycle
    }

    // ListNode class (inner class for linked list problems)
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }



    // 2. Find the Start of a Linked List Cycle
    //   - Real-world application: Detecting the beginning of a loop in a circular buffer
    //     used in real-time data processing, allowing for resetting the buffer.
    //   - Problem: Given the head of a linked list with a cycle, find the node where the cycle begins.
    //   - Time Complexity: O(n)
    //   - Space Complexity: O(1)
    public static ListNode findCycleStart(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }
        ListNode slow = head;
        ListNode fast = head;

        // Detect the cycle
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                break; // Cycle detected, break the loop
            }
        }

        // If no cycle, return null
        if (fast == null || fast.next == null) {
            return null;
        }
        // Find the start of the cycle
        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }
        return slow; // Returns the start of the cycle
    }


    // 3. Happy Number
    //   - Real-world application: Used in software testing to detect infinite loops
    //     in state transition systems. For example, testing if a configuration setting
    //     repeatedly leads back to itself without reaching a terminal state.
    //   - Problem: A happy number is a number which eventually reaches 1 when replaced by the sum of the square of each digit.
    //     For example, 13 is a happy number because 1^2 + 3^2 = 10, 1^2 + 0^2 = 1.
    //     Determine if a number is a happy number.
    //   - Time Complexity: O(log n)
    //   - Space Complexity: O(1)
    public static boolean isHappy(int n) {
        int slow = n;
        int fast = n;

        do {
            slow = findSquareSum(slow);
            fast = findSquareSum(findSquareSum(fast));
        } while (slow != fast);

        return slow == 1;
    }

    private static int findSquareSum(int num) {
        int sum = 0;
        while (num > 0) {
            int digit = num % 10;
            sum += digit * digit;
            num /= 10;
        }
        return sum;
    }


    // 4. Middle of the Linked List
    //  - Real-world application: Finding the middle point in a linked list of tasks
    //    in a task management system, which can be useful for dividing tasks or
    //    implementing a "jump to middle" feature.
    //  - Problem: Given the head of a singly linked list, return the middle node of the linked list.
    //    If there are two middle nodes, return the second middle node.
    //  - Time Complexity: O(n)
    //  - Space Complexity: O(1)

    public static ListNode middleNode(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }


    // 5. Circular Array Loop
    // - Real-world application: Detecting cyclic dependencies in a configuration file,
    //   where each element represents a jump instruction to another element. This can
    //   help prevent infinite loops during configuration loading or processing.
    // - Problem: You are given a circular array nums of n integers, where each number nums[i] is the number of steps
    //   you take forward or backward starting from index i.
    //   Return true if there is a cycle in nums, otherwise return false.
    // - Time Complexity: O(n)
    // - Space Complexity: O(1)
    public static boolean hasCircularArrayLoop(int[] nums) {
        if (nums == null || nums.length < 2) {
            return false;
        }

        int n = nums.length;
        for (int i = 0; i < n; i++) {
            int slow = i, fast = i;
            boolean isForward = nums[i] > 0; // Store the direction

            while (true) {
                // Move slow pointer
                slow = nextStep(nums, slow, isForward);
                // Move fast pointer twice
                fast = nextStep(nums, fast, isForward);
                if (fast != -1) { // Check for -1 before nextStep
                    fast = nextStep(nums, fast, isForward);
                }

                if (slow == -1 || fast == -1) {
                    break; // No cycle found for this starting point
                }

                if (slow == fast) {
                    // Check for non-trivial cycle (length > 1)
                    if (slow == nextStep(nums, slow, isForward)) {
                        break; // Cycle of length 1, invalid
                    }
                    return true; // Cycle found
                }
            }
        }
        return false;
    }

    private static int nextStep(int[] nums, int index, boolean prevIsForward) {
        int n = nums.length;
        if (index < 0 || index >= n) return -1;

        boolean currentIsForward = nums[index] > 0;
        if (currentIsForward != prevIsForward) {
            return -1; // Different direction, not a valid step
        }

        int nextIndex = (index + nums[index]) % n;
        if (nextIndex < 0) {
            nextIndex += n; // Handle negative indices in circular array
        }
        return nextIndex;
    }



    public static void main(String[] args) {
        // 1. Linked List Cycle Detection
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(3);
        head1.next.next.next = new ListNode(4);
        head1.next.next.next.next = head1.next; // Create a cycle
        System.out.println("Linked List Cycle Detection: " + hasCycle(head1)); // Output: true

        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(2);
        head2.next.next = new ListNode(3);
        System.out.println("Linked List Cycle Detection: " + hasCycle(head2)); // Output: false

        // 2. Find the Start of a Linked List Cycle
        ListNode cycleStart = findCycleStart(head1);
        if (cycleStart != null) {
            System.out.println("Cycle Start Node Value: " + cycleStart.val); // Output: 2
        } else {
            System.out.println("No cycle found.");
        }

        ListNode head3 = new ListNode(1);
        head3.next = new ListNode(2);
        head3.next.next = new ListNode(3);
        System.out.println("Cycle Start of list without cycle: " + findCycleStart(head3)); //null

        // 3. Happy Number
        System.out.println("Is 13 a happy number? " + isHappy(13)); // Output: true
        System.out.println("Is 4 a happy number? " + isHappy(4));   // Output: false

        // 4. Middle of the Linked List
        ListNode head4 = new ListNode(1);
        head4.next = new ListNode(2);
        head4.next.next = new ListNode(3);
        head4.next.next.next = new ListNode(4);
        head4.next.next.next.next = new ListNode(5);
        System.out.println("Middle Node: " + middleNode(head4).val); // Output: 3

        ListNode head5 = new ListNode(1);
        head5.next = new ListNode(2);
        head5.next.next = new ListNode(3);
        head5.next.next.next = new ListNode(4);
        head5.next.next.next.next = new ListNode(5);
        head5.next.next.next.next.next = new ListNode(6);
        System.out.println("Middle Node of even length list: " + middleNode(head5).val); //4

        // 5. Circular Array Loop
        int[] nums1 = {2, -1, 1, 2, 2};
        System.out.println("Circular Array Loop: " + hasCircularArrayLoop(nums1)); // Output: true

        int[] nums2 = {-2, 1, -1, -2, -2};
        System.out.println("Circular Array Loop: " + hasCircularArrayLoop(nums2)); // Output: false

        int[] nums3 = {1, 2};
        System.out.println("Circular Array Loop: " + hasCircularArrayLoop(nums3)); // false

        int[] nums4 = {2, -1, 1, -2, -2};
        System.out.println("Circular Array Loop 4: " + hasCircularArrayLoop(nums4)); // false

        int[] nums5 = {1, 2, -1, 2, 2};
        System.out.println("Circular Array Loop 5: " + hasCircularArrayLoop(nums5)); // true
    }
}

